### DISCUSSION TITLE
Welcome to BlackSheep Discussions!

### DISCUSSION THREAD

**Discussion by RobertoPrevato** [Neoteroi/BlackSheep#57] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by 4l1fe** [comment]
Yet another one ASGI framework, why?

**Reply by RobertoPrevato** [reply]
Because I initially included an HTTP Server implementation in my framework, using Cython to implement the response handling part.
My work proved to be good since [benchmarks from TechEmpower](https://www.techempower.com/benchmarks/#section=test&runid=76a34044-54d6-4349-adfe-863c2d5ae756&hw=ph&test=db&l=z8kflr-v&a=2) showed that my framework is one of the fastest for Python today:

![image](https://user-images.githubusercontent.com/2576032/104302572-3f2cea80-54c9-11eb-9534-5dfb5cae3bc7.png)

However, maintaining an HTTP Server implementation together with the implementation of a web framework is a lot of work. I exchanged a few emails with Tom Christie and he recommended me to drop my HTTP Server implementation and make my web framework ASGI compatible.
Tom Christie's advice was good because I can dedicate myself on higher level abstractions rather than lower level HTTP handling details, like implementing support for HTTP2 and HTTP3.

If I will ever work on an HTTP Server again I will release it anyway as a separate component, but I don't see this happening unless my work will be sponsored. Because I don't have time and energy to work on something more for free.

---

On the other hand, I wanted my web framework to support [dependency injection out of the box](https://www.neoteroi.dev/blacksheep/dependency-injection/) and [request handler normalization](https://www.neoteroi.dev/blacksheep/requests), like ASP.NET Core does - because in my opinion this is an important feature in a web framework. Tom Christie has a different point of view on this matter, so I prefer to keep my own web framework on top of uvicorn rather than abandoning my project.

I wrote more on this subject here [on my blog](https://robertoprevato.github.io/Presenting-BlackSheep/).

**Comment by skivis** [comment]
Just found BlackSheep an hour ago, been reading the docs & examples and browsing through the repo, and I have to say I love it and I'm super stoked about this project. Definitely gonna give it a go and I hope it gains some more traction ASAP!

**Reply by RobertoPrevato** [reply]
Thank You, @skivis! It means a lot to me. I hope you will find it useful. I dedicated lots of time to study and work on this project. In the last days I took a short break but I will soon start working on more improvements.
